<div class="row">
  <div class="col-md-12">
    <table class="table table-striped">
      <tr> <td>13.00 - 13.10</td>  <td> <strong>Welcome</strong> <a href="https://github.com/ScienceParkStudyGroup/2019-03-07-rnaseq-workshop/blob/gh-pages/pages/00.welcome.md">Workshop presentation</a></td> <td>Marc</td></tr>
      <tr> <td>13.10 - 13.20</td>  <td> <strong>Introduction</strong> <a href="https://github.com/ScienceParkStudyGroup/2019-03-07-rnaseq-workshop/blob/gh-pages/pages/01-rnaseq-experimental-design.md">01. Experimental design and Next-Generation Sequencing for RNA-Seq</a></td> <td>Marc</td></tr>
      <tr> <td>13.30 - 14.00</td>  <td> <strong>Introduction</strong> <a href="#"></a>02. RNA-Seq post-sequencing steps</td> <td>Ernest</td> </tr>
      <tr> <td>14.00 - 14.30</td>  <td> <strong>Introduction</strong> 03. Set-up: fastq files, genome sequence, annotation, softwares</td> <td>Tijs</td></tr>
      <tr> <td>14.30 - 15.00</td>  <td> <strong>Practical</strong> 04. QC and trimming of raw reads using <em>fastQC</em> and <em>Trimmomatic</em></td> <td>Tijs</td> </tr>
      <tr> <td>15:00 - 15.15</td>  <td>Coffee break</td> <td> </td></tr>
      <tr> <td>15.15 - 16.00</td>  <td> <strong>Practical</strong> 05. Read alignment with <em>hisat2</em> and counting with <em>featureCounts</em> </td> <td>Tijs</td> </tr>
      <tr> <td>16.00 - 16.15</td>  <td> <strong>Practical</strong> 06. File preparation for use in the <em>ideal</em> online application</td> <td>Marc</td> </tr>
      <tr> <td>16:15 - 17.00</td>  <td> <strong>Practical</strong> 07. Data exploration using the <em>ideal</em> Shiny app</td> <td>Marc</td> </tr>
    </table>
  </div>
</div>
